File: components/esp_board_manager/peripherals/periph_ldo/periph_ldo.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# LDO peripheral config parser
VERSION = 'v1.0.0'

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'generators'))

def get_includes() -> list:
    """Return list of required include headers for LDO peripheral"""
    return [
        'esp_ldo_regulator.h',
    ]

def parse(name: str, config: dict) -> dict:
    """Parse LDO peripheral configuration from YAML to C structure

    Args:
        name: Peripheral name
        config: Configuration dictionary from YAML

    Returns:
        Dictionary containing LDO configuration structure for esp_ldo_channel_config_t
    """
    try:
        # Get the config dictionary
        ldo_cfg = config.get('config', {})

        # Get ldo_cfg with validation
        chan_id = ldo_cfg.get('chan_id', 3)
        if not isinstance(chan_id, int) or chan_id < 0:
            raise ValueError(f'Invalid chan_id: {chan_id}. Must be an integer >= 0.')
        voltage_mv = ldo_cfg.get('voltage_mv', 2500)
        if not isinstance(voltage_mv, int) or voltage_mv < 0:
            raise ValueError(f'Invalid voltage_mv: {voltage_mv}. Must be an integer >= 0.')
        adjustable = ldo_cfg.get('adjustable', 1)
        if adjustable not in [0, 1]:
            raise ValueError(f'Invalid adjustable: {adjustable}. Must be 0 or 1.')
        owned_by_hw = ldo_cfg.get('owned_by_hw', 0)
        if owned_by_hw not in [0, 1]:
            raise ValueError(f'Invalid owned_by_hw: {owned_by_hw}. Must be 0 or 1.')

        # Create base config
        result = {
            'struct_type': 'esp_ldo_channel_config_t',
            'struct_var': f'{name}_cfg',
            'struct_init': {
                'chan_id': chan_id,
                'voltage_mv': voltage_mv,
                'flags': {
                    'adjustable': adjustable,
                    'owned_by_hw': owned_by_hw,
                }
            }
        }

        return result

    except ValueError as e:
        # Re-raise ValueError with more context
        raise ValueError(f"YAML validation error in LDO peripheral '{name}': {e}")
    except Exception as e:
        # Catch any other exceptions and provide context
        raise ValueError(f"Error parsing LDO peripheral '{name}': {e}")

File: components/esp_board_manager/peripherals/periph_ldo/periph_ldo.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include "esp_err.h"
#include "esp_log.h"
#include "periph_ldo.h"
#include "esp_ldo_regulator.h"

static const char *TAG = "PERIPH_LDO";

int periph_ldo_init(void *cfg, int cfg_size, void **periph_handle)
{
    if (!cfg || !periph_handle || cfg_size != sizeof(esp_ldo_channel_config_t)) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }

    esp_ldo_channel_handle_t ldo_phy_chan = NULL;
    esp_err_t err = esp_ldo_acquire_channel((esp_ldo_channel_config_t *)cfg, &ldo_phy_chan);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "LDO acquire channel failed: %s", esp_err_to_name(err));
        return -1;
    }

    ESP_LOGI(TAG, "LDO initialize success");
    *periph_handle = ldo_phy_chan;
    return 0;
}

int periph_ldo_deinit(void *periph_handle)
{
    if (!periph_handle) {
        ESP_LOGE(TAG, "Invalid handle");
        return -1;
    }
    ESP_LOGI(TAG, "LDO deinitialize");
    esp_ldo_channel_handle_t handle = (esp_ldo_channel_handle_t)periph_handle;
    esp_err_t err = esp_ldo_release_channel(handle);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "LDO del failed: %s", esp_err_to_name(err));
        return -1;
    }
    return 0;
}
